Dockerfile

```diff
@@ -1,201 +1,98 @@
-ENV LD_LIBRARY_PATH=$JAVA_HOME/lib:$JAVA_HOME/lib/server
-ENV USER=localstack
-ENV PYTHONUNBUFFERED=1
+RUN apt-get update && \
+    apt-get install -y --no-install-recommends \
+        ca-certificates curl gnupg git make openssl tar pixz zip unzip groff-base iputils-ping \
+        && apt-get upgrade -y libexpat1 \
+    && apt-get clean && rm -rf /var/lib/apt/lists/*
+
+# Install Node.js without unnecessary dependencies
+RUN dpkgArch="$(dpkg --print-architecture)" && \
+    case "${dpkgArch##*-}" in \
+        amd64) ARCH='x64';; \
+        arm64) ARCH='arm64';; \
+        *) echo "unsupported architecture"; exit 1 ;; \
+    esac && \
+    curl -fsSL https://deb.nodesource.com/setup_18.x | bash - && \
+    apt-get install -y nodejs && \
+    npm cache clean --force
+
+# Set workdir and create localstack user
+WORKDIR /opt/code/localstack
+RUN useradd -ms /bin/bash localstack && \
+    mkdir -p /var/lib/localstack /tmp/localstack && \
+    chmod 777 /var/lib/localstack /tmp/localstack && \
+    touch /tmp/localstack/.marker
+
+# Install basic global tools
+RUN pip install --no-cache-dir --upgrade virtualenv
+
+# Install entrypoint script
+COPY bin/docker-entrypoint.sh /usr/local/bin/
+COPY bin/hosts /etc/hosts
+
+# Environment variables
+ENV USER=localstack LANG=C.UTF-8 PYTHONUNBUFFERED=1
 
 # Install the latest version of awslocal globally
-RUN --mount=type=cache,target=/root/.cache \
-    pip3 install --upgrade awscli awscli-local requests
+RUN pip install --no-cache-dir --upgrade awscli awscli-local requests
 
-
-#
-# builder: Stage which installs the dependencies of LocalStack Community
-#
+# Builder stage: Install dependencies for LocalStack Community
 FROM base AS builder
-ARG TARGETARCH
-
-# Install build dependencies to base
-RUN --mount=type=cache,target=/var/cache/apt \
-    apt-get update && \
-        # Install dependencies to add additional repos
-        # g++ is a workaround to fix the JPype1 compile error on ARM Linux "gcc: fatal error: cannot execute ‘cc1plus’"
-        apt-get install -y gcc g++
+RUN apt-get update && \
+    apt-get install -y --no-install-recommends gcc g++ && \
+    apt-get clean && rm -rf /var/lib/apt/lists/*
 
-# upgrade python build tools
-RUN --mount=type=cache,target=/root/.cache \
-    (virtualenv .venv && . .venv/bin/activate && pip3 install --upgrade pip wheel setuptools)
+# Upgrade Python build tools
+RUN virtualenv .venv && . .venv/bin/activate && pip install --upgrade pip wheel setuptools
 
-# add files necessary to install runtime dependencies
-ADD Makefile pyproject.toml requirements-runtime.txt ./
-# add the localstack start scripts (necessary for the installation of the runtime dependencies, i.e. `pip install -e .`)
-ADD bin/localstack bin/localstack.bat bin/localstack-supervisor bin/
+# Add necessary files and install runtime dependencies
+COPY Makefile pyproject.toml requirements-runtime.txt ./
+COPY bin/localstack bin/localstack.bat bin/localstack-supervisor bin/
```

Comment:
+1, favoring add over copy [has been suggested by Docker due to its transparency / security / simplicity](https://www.docker.com/blog/docker-best-practices-understanding-the-differences-between-add-and-copy-instructions-in-dockerfiles/).